src/puppetlabs/puppetdb/catalogs.clj

```diff
@@ -367,13 +367,32 @@
            (throw (IllegalArgumentException. (format "Catalog must be specified as a string or a map, not '%s'" (class catalog)))))))
 
 (defmethod parse-catalog String
-  [catalog version]
+  [catalog version received-time]
   {:pre   [(string? catalog)]
    :post  [(map? %)]}
-  (parse-catalog (json/parse-string catalog true) version))
+  (parse-catalog (json/parse-string catalog true) version received-time))
+
+(defmethod parse-catalog 4
+  [catalog version received-time]
+  {:pre [(map? catalog)
+         (number? version)]
+   :post [(map? %)]}
+  (-> catalog
+      (assoc :producer_timestamp received-time)
```

Comment:
Not sure exactly what you're asking. I'm using received-time for producer-timestamp only for v4 of the catalog wire format. v5 and v6 should use the producer-timestamp included in the message. We already do this via our SQL migration of old catalogs when users upgrade, this is just handling old stuff that has not yet been persisted. It should have the same end result as the SQL changes.